### PR TITLE
fix affine_transform (fixes #1547 and #5702)

### LINKED ISSUE #1547: Inconsistent use of offset in ndimage.interpolation.affine_transform() (Trac #1020)
_Original ticket http://projects.scipy.org/scipy/ticket/1020 on 2009-10-14 by trac user jvrantwijk, assigned to unknown._

I am confused about the meaning of the "offset" parameter in scipy.ndimage.interpolation.affine_transform(). It seems to apply the offset before applying the scaling matrix, but only in case the scaling matrix is specified as a 1-D array.

The following example makes this more clear.
My test array looks like this:

    >>> w
    array([[ 0.,  1.,  2.,  3.],
           [ 4.,  5.,  6.,  7.]])

On this array, I run an affine transform scaling the second index by 0.5. If I do this without offset, it works just like I would expect:

    >>> from scipy.ndimage.interpolation import affine_transform
    >>> affine_transform(w, matrix=(1,0.5), offset=(0,0), order=1)
    array([[ 0. ,  0.5,  1. ,  1.5],
           [ 4. ,  4.5,  5. ,  5.5]])

In the next test I add an offset of 1 to the second index. I would expect this offset to apply in the index space of w[], which means that I expect the number 1 in the top left corner of the result.
However:

    >>> affine_transform(w, matrix=(1,0.5), offset=(0,1), order=1)
    array([[ 0.5,  1. ,  1.5,  2. ],
           [ 4.5,  5. ,  5.5,  6. ]])

Ok, so apparently it applied the offset to the index before scaling the index space. Fine. It seems counterintuitive, but fine.

However, now I try the same thing but specifying the full 2x2 matrix instead of just the diagonal entries:

    >>> affine_transform(w, matrix=[[1,0],[0,0.5]], offset=(0,1), order=1)
    array([[ 1. ,  1.5,  2. ,  2.5],
           [ 5. ,  5.5,  6. ,  6.5]])

And suddenly now the offset DOES work like I expected.

This is inconsistent. I think affine_transform() is simply wrong in the diagonal-matrix case.

### PR DESCRIPTION
**Pull request by burnpanck**

### CODE REVIEW

File: scipy/ndimage/interpolation.py [hunk 1]
```diff
@@ -339,6 +339,20 @@ def affine_transform(input, matrix, offset=0.0, output_shape=None,
...
     mode.
 
+    Given an output image pixel index vector ``o``, the pixel value
+    is determined from the input image at position ``np.dot(matrix,o) + offset``.
+
+    A diagonal matrix can be specified by supplying a one-dimensional
+    array-like to the matrix parameter, in which case a more efficient
+    algorithm is applied.
+
+    .. versionchanged:: 0.18.0
```
Comment (rgommers) [inline scipy/ndimage/interpolation.py]: Is this a thing? The refguide check fails on TravisCI, probably on this.

**Comment by rgommers** [comment]
Thanks for picking this up @burnpanck.

**Comment by rgommers** [comment]
Would be good to mention this in the 0.18.0 release notes (in ``doc/release/``) under "backwards incompatible changes".

File: scipy/ndimage/interpolation.py (hunk 1 above)
Comment (burnpanck) [inline scipy/ndimage/interpolation.py]: I actually don't know... Do you think it is the `.. versionchanged::` that makes it fail? I found that in the spinx documentation, but I really don't know sphinx nor scipy's documentation configuration at all... I'll try to investigate locally.

Comment (pv) [inline scipy/ndimage/interpolation.py]: It's a thing, but the refguide checker doesn't understand it since it's not really sphinx.

**Comment by pv** [comment]
thanks, merged

**Pull request merged**